src/main/java/org/cloudname/uritemplate/UriTemplate.java

```diff
@@ -118,6 +151,31 @@ private static String expandExpression(String expression, Map<String,String> var
     }
 
     /**
+     * Extract all variables from a given expression.
+     *
+     * @param expression to parse
+     * @return {@link Set} of all variables in the expression
+     */
+    private static Set<String> extractVarsFromExpression(String expression) {
+        Character operator = expression.charAt(0); // the first character may be an operator
+        int varspecStart = 0; // the varspec tentatively starts here
+        Set<String> extractedVars = new HashSet<String>();
+        if (params.containsKey(operator)) {
+            // found an operator we know about, so skip it to find the varspec
+            varspecStart += 1;
+        }
+
+        // look up the algorithm parameters, and pick apart the varspec
+        String varspecString = expression.substring(varspecStart);
+        String[] varspec = varspecString.split(",");
+
+        for (String var : varspec) {
+            extractedVars.add(var);
+        }
+        return extractedVars;
+    }
+
+    /**
```

Comment:
Should we have a method that checks that a HashMap contains all used variables as well? Then this could be used directly to test if the template "would resolve".